File: 0539-minimum-time-difference/0539-minimum-time-difference.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

int convertToMinutes(char* time) {
    int hours, minutes;
    sscanf(time, "%d:%d", &hours, &minutes);
    return hours * 60 + minutes;
}

int compare(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int findMinDifference(char** timePoints, int timePointsSize) {
    // Convert all time strings to minutes
    int* minutes = (int*)malloc(timePointsSize * sizeof(int));
    
    for (int i = 0; i < timePointsSize; i++) {
        minutes[i] = convertToMinutes(timePoints[i]);
    }
    
    // Sort the time points in ascending order
    qsort(minutes, timePointsSize, sizeof(int), compare);
    
    int minDiff = INT_MAX;
    
    // Calculate the minimum difference between consecutive time points
    for (int i = 1; i < timePointsSize; i++) {
        int diff = minutes[i] - minutes[i - 1];
        if (diff < minDiff) {
            minDiff = diff;
        }
    }
    
    // Check the difference between the first and last time points (wrap-around case)
    int wrapAroundDiff = (1440 - minutes[timePointsSize - 1] + minutes[0]);
    if (wrapAroundDiff < minDiff) {
        minDiff = wrapAroundDiff;
    }
    
    free(minutes);
    
    return minDiff;
}

File: 0539-minimum-time-difference/0539-minimum-time-difference.go
package main

import (
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
)

func convertToMinutes(time string) int {
	parts := strings.Split(time, ":")
	hours, _ := strconv.Atoi(parts[0])
	minutes, _ := strconv.Atoi(parts[1])
	return hours*60 + minutes
}

func findMinDifference(timePoints []string) int {
	// Convert all time strings to minutes
	minutes := make([]int, len(timePoints))
	for i, time := range timePoints {
		minutes[i] = convertToMinutes(time)
	}

	// Sort the time points in ascending order
	sort.Ints(minutes)

	// Initialize the minimum difference to a large number
	minDiff := math.MaxInt32

	// Calculate the minimum difference between consecutive time points
	for i := 1; i < len(minutes); i++ {
		diff := minutes[i] - minutes[i-1]
		if diff < minDiff {
			minDiff = diff
		}
	}

	// Check the difference between the first and last time points (wrap-around case)
	wrapAroundDiff := 1440 - minutes[len(minutes)-1] + minutes[0]
	if wrapAroundDiff < minDiff {
		minDiff = wrapAroundDiff
	}

	return minDiff
}